src/plugins/discover/public/application/components/source_viewer/source_viewer.tsx

```diff
@@ -0,0 +1,122 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import React, { useEffect, useState } from 'react';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { monaco } from '@kbn/monaco';
+import { EuiEmptyPrompt, EuiLoadingSpinner, EuiSpacer, EuiText } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { useEsDocSearch } from '../doc/use_es_doc_search';
+import { JSONCodeEditorCommonMemoized } from '../json_code_editor/json_code_editor_common';
+import { ElasticRequestState } from '../doc/elastic_request_state';
+import { getServices } from '../../../../public/kibana_services';
+import { SEARCH_FIELDS_FROM_SOURCE } from '../../../../common';
+
+interface SourceViewerProps {
+  id: string;
+  index: string;
+  indexPatternId: string;
+  hasLineNumbers: boolean;
+  width?: number;
+}
+
+export const SourceViewer = ({
+  id,
+  index,
+  indexPatternId,
+  width,
+  hasLineNumbers,
+}: SourceViewerProps) => {
+  const [editor, setEditor] = useState<monaco.editor.IStandaloneCodeEditor>();
+  const [jsonValue, setJsonValue] = useState<string>('');
+  const indexPatternService = getServices().data.indexPatterns;
+  const useNewFieldsApi = !getServices().uiSettings.get(SEARCH_FIELDS_FROM_SOURCE);
+  const [reqState, hit] = useEsDocSearch({
+    id,
+    index,
+    indexPatternId,
+    indexPatternService,
+    requestSource: useNewFieldsApi,
+  });
+
+  useEffect(() => {
+    if (reqState === ElasticRequestState.Found && hit) {
+      setJsonValue(JSON.stringify(hit, undefined, 2));
+    }
+  }, [reqState, hit]);
+
+  // setting editor height based on lines height and count to stretch and fit its content
+  useEffect(() => {
+    if (!editor) {
+      return;
+    }
+    const editorElement = editor.getDomNode();
+
+    if (!editorElement) {
+      return;
+    }
+
+    const lineHeight = editor.getOption(monaco.editor.EditorOption.lineHeight);
+    const lineCount = editor.getModel()?.getLineCount() || 1;
+    const height = editor.getTopForLineNumber(lineCount + 1) + lineHeight;
+    if (!jsonValue || jsonValue === '') {
+      editorElement.style.height = '0px';
+    } else {
+      editorElement.style.height = `${height}px`;
+    }
+    editor.layout();
+  }, [editor, jsonValue]);
+
+  const loadingState = (
+    <div className="euiDataGrid__loading">
+      <EuiText size="xs" color="subdued">
+        <EuiLoadingSpinner />
+        <EuiSpacer size="s" />
+        <FormattedMessage id="discover.loadingJSON" defaultMessage="Loading JSON" />
+      </EuiText>
+    </div>
+  );
+
+  const errorMessageTitle = (
+    <h2>
+      {i18n.translate('discover.sourceViewer.errorMessageTitle', {
+        defaultMessage: 'An Error Occurred',
+      })}
+    </h2>
+  );
+  const errorMessage = (
+    <p>
+      {i18n.translate('discover.sourceViewer.errorMessage', {
+        defaultMessage: 'Could not fetch data at this time. Refresh the tab to try again.',
```

Comment:
I wonder if `Refresh the tab to try again.` could be improved in terms of wording, since there is no good way to refresh, except clicking on the tab again. Reloading the browser tab would mean to loose the open doc in the main view. So I'd suggest to replace by `Refresh data.` and to link it so the user could try again to fetch the data.

![Bildschirmfoto 2021-06-22 um 11 04 12](https://user-images.githubusercontent.com/463851/122897634-77730780-d34a-11eb-8a3c-5574db8ed43a.png)

Reply:
thx for adapting, works well!